src/core/tests/queryObserver.test.tsx

```diff
@@ -490,4 +495,275 @@ describe('queryObserver', () => {
+    const placeholderData1 = { value: 'data' }
+    const placeholderData2 = { value: 'data' }
+
+    const observer = new QueryObserver(queryClient, {
+      queryKey: key,
+      queryFn: () => data,
+      select: () => data,
+    })
+
+    observer.setOptions({
+      queryKey: key,
+      queryFn: () => data,
+      select: () => {
+        if (true) return selectedData1
+      },
+      placeholderData: placeholderData1,
+    })
+
+    observer.setOptions({
+      queryKey: key,
+      queryFn: () => data,
+      select: () => {
+        return selectedData2
+      },
+      placeholderData: placeholderData2,
+      structuralSharing: false,
+    })
+
+    expect(observer.getCurrentResult().data).toBe(selectedData2)
+  })
+
+  test('should not use an undefined value returned by select as placeholderdata', () => {
+    const key = queryKey()
+
+    const data = { value: 'data' }
+    const selectedData = { value: 'data' }
+    const placeholderData1 = { value: 'data' }
+    const placeholderData2 = { value: 'data' }
+
+    const observer = new QueryObserver(queryClient, {
+      queryKey: key,
+      queryFn: () => data,
+      select: () => data,
+    })
+
+    observer.setOptions({
+      queryKey: key,
+      queryFn: () => data,
+      select: () => {
+        if (true) return selectedData
+      },
+      placeholderData: placeholderData1,
+    })
+
+    expect(observer.getCurrentResult().isPlaceholderData).toBe(true)
+
+    observer.setOptions({
+      queryKey: key,
+      queryFn: () => data,
+      //@ts-ignore
+      select: () => undefined,
+      placeholderData: placeholderData2,
+    })
+
+    expect(observer.getCurrentResult().isPlaceholderData).toBe(false)
+  })
+
+  test('updateResult should not notify cache listeners if cache option is false', async () => {
+    const key = queryKey()
+
+    const data1 = { value: 'data 1' }
+    const data2 = { value: 'data 2' }
+
+    await queryClient.prefetchQuery(key, () => data1)
+    const observer = new QueryObserver(queryClient, {
+      queryKey: key,
+    })
+    await queryClient.prefetchQuery(key, () => data2)
+
+    const spy = jest.fn()
+    const unsubscribe = queryClient.getQueryCache().subscribe(spy)
+    observer.updateResult({ cache: false })
+
+    expect(spy).toHaveBeenCalledTimes(0)
+
+    unsubscribe()
+  })
+
+  test('should not notify observer when the stale timeout expires and the current result is stale', async () => {
+    const key = queryKey()
+    const queryFn = () => 'data'
+
+    await queryClient.prefetchQuery(key, queryFn)
+    const observer = new QueryObserver(queryClient, {
+      queryKey: key,
+      queryFn,
+      staleTime: 20,
+    })
+
+    const spy = jest.fn()
+    const unsubscribe = observer.subscribe(spy)
+    await queryClient.refetchQueries(key)
+    await sleep(10)
+
+    // Force isStale to false
+    // because no use case has been found to reproduce this condition
+    // @ts-ignore
+    observer['currentResult'].isStale = true
```

Comment:
hmm, the `staleTime` is `20`, but we only wait for `10`. I think if we reverse this (set staleTime to 10 and wait for 20), `isStale` should automatically go to `true`.

Reply:
If I do this change the else in the block of code below is not covered anymore:
![Uploading Screenshot 2021-11-06 at 20.56.15.png…]()

As mentioned in the test comment I have not found a use case for this condtion in the `setTimeout`:
https://github.com/tannerlinsley/react-query/blob/b52b2779b32b5cf0c21733e7d86b5882b8a611a6/src/core/queryObserver.ts#L364-L368

Reply:
I think what we'd need to do is manually set the query to stale by calling `queryClient.invalidateQueries(queryKey, { refetchActive: false })`. This would mark the query as `isInvalidated`, so `isStale` will already be true on the currentResult when the timer fires:

https://github.com/tannerlinsley/react-query/blob/e7a3207f74b27bc112d6ab583831cceac60f94c5/src/core/query.ts#L288-L294

Reply:
Tried but the `else` is still not covered and even the timeout handler is not called at all in this case.
It's because:  
`query.invalidate()`  => `dispatch()` => `onQueryUpdate()` => `updateTimers()` => `updateStaleTimeout()` => `clearStaleTimeout()`
That's why I don't see a use case for this code coverage.
It looks like a useless condition in the `setTimeout()` handler (or a bug smell?).

Reply:
True, it could be that the timer is reset once the query is marked as stale by invalidation. Maybe the condition is unnecessary but I’d rather not remove it now. Do you think it makes sense to keep the test?

Reply:
I was wondering the same for this kind of tests that test something that should never happen..
We can keep this test to improve the code coverage and I use the comment prefix `// Force ... ` to identify this kind of situation. 
So it makes sense to keep this test, it's a kind a code documentation about the tested condition.